Updates to pre-built Security ML jobs

ajosh0504 opened the pull request:
## Summary
This PR makes the following updates to our pre-built Security ML jobs:
- Adds user-friendly names to our pre-built Anomaly Detection jobs. These will be displayed in the Anomalies tab on the new Entity Analytics page in the Security App instead of job IDs.
- Fixes formatting
- One job was missing the `security` job group which is required to display jobs in the Security App. Added that as well.
- Changed the names of two modules: `siem_cloudtrail` -> `security_cloudtrail` and `siem_packetbeat` -> `security_packetbeat`.  This should have happened a while ago per [this](https://github.com/elastic/kibana/issues/69319) issue.

## Side effects
- Any QA tests that reference the `siem_cloudtrail` and `siem_packetbeat` modules will need to be changed to reference the new modules instead 
- Any references to the siem modules in the Security App will need to be updated

elasticmachine commented:
Pinging @elastic/ml-ui (:ml)

ajosh0504 commented:
@machadoum Any thoughts on backporting? Or do we want to skip?

sophiec20 commented:
Please keep on main - changes are too significant for backporting, as FF has passed.

ajosh0504 commented:
@sophiec20 Yeah we didn't intend for this to go in in 8.6. Fair point about keeping it on main owing to the magnitude of changes made.

peteharverson commented on x-pack/plugins/ml/server/models/data_recognizer/modules/security_linux/ml/v3_linux_rare_metadata_process.json:
```diff
@@ -40,6 +40,8 @@
-  },
-    "created_by": "ml-module-security-linux"  }
+    },
+    "created_by": "ml-module-security-linux",
```
Some of the jobs in this module use a `created_by` of `ml-module-security-linux` and some use `ml-module-security-linux-v3`. It's been like this for a while - is this difference deliberate?

ajosh0504 replied:
This has been around for a while, but I can make updates to keep things consistent. And if both have been working so far, I don't think changing to either one will break things. I'm leaning towards making that field version independent.

peteharverson commented on x-pack/plugins/ml/server/models/data_recognizer/modules/security_windows/ml/v3_windows_anomalous_script.json:
```diff
@@ -48,6 +49,7 @@
       }
-    ]
+    ],
+    "security_app_display_name": "Suspicious Powershell Script"
```
This job is missing the `created_by` field.

peteharverson commented on x-pack/plugins/ml/server/models/data_recognizer/modules/security_windows/ml/v3_windows_anomalous_service.json:
```diff
@@ -32,19 +32,20 @@
       "updated_date": "5/16/2022"
-  },
+    },
     "created_by": "ml-module-security-windows-v3",
```
As with the linux jobs, some of the jobs in this module have `-v3` at the end of the `created_by` and some don't. Is this deliberate? I'd vote for them being consistent.

ajosh0504 replied:
Agreed.

peteharverson commented:
@ajosh0504 if you keep the changes to the module IDs in this PR, you'll need to adjust some of the API tests in `x-pack/test/api_integration/apis/ml/modules` which reference the old module IDs. Please reach out if you need help finding what to change. I think the 3 files that need some small edits are `recognize_module.ts`, `setup_module.ts` and `get_module.ts`.

machadoum commented:
> Two `siem` module IDs have been renamed: `siem_packetbeat` -> `security_packetbeat` and `siem_cloudtrail` -> `security_cloudtrail`...
> ... I guess those IDs might need updating somewhere in the Security app

I searched for the IDs inside the security repo, and this is the only reference that I could find:
https://github.com/elastic/kibana/blob/f85c39e5f66f5e770a611ff3eae2f4ecfab5d7b3/x-pack/plugins/security_solution/public/common/components/ml_popover/ml_modules.tsx#L14-L15

ajosh0504 commented:
@droberts195 Thanks for catching that. I've updated the PR title, as well as the description to indicate that multiple changes are being made to the pre-built ML jobs in this PR. I've also called out that updates will need to be made in a few other places and I'll try to update as many tests as possible in this PR per recommendations from @peteharverson and @machadoum.

ajosh0504 commented:
@peteharverson I updated the test files you mentioned above. While updating those, I realized that there's some other siem references in the tests owing to the folders [here](https://github.com/elastic/kibana/tree/main/x-pack/test/functional/es_archives/ml) being named as siem_*. Is this sometime we want to modify in this PR as well?

Another thing I noted was references to moduleIDs containing versions (for example [here](https://github.com/elastic/kibana/blob/main/x-pack/test/api_integration/apis/ml/modules/get_module.ts#L30)), but those module IDs don't really exist [here](https://github.com/elastic/kibana/tree/main/x-pack/plugins/ml/server/models/data_recognizer/modules)?

jpdjere reviewed (approved):
Codeowners review from Detection-Rules LGTM

jpdjere commented:
@rylnd Could you double check if the renaming of the modules:
```
siem_cloudtrail -> security_cloudtrail
siem_packetbeat -> security_packetbeat
```

has any impact on your side? 

We have them hardcoded on `x-pack/plugins/security_solution/public/common/components/ml_popover/ml_modules.tsx` and this PR modifies that, but these are compared in the `useSecurityJobs` hook against the modules fetched by the `/api/ml/modules/get_module/{moduleId}` API, which I'm not familiar with.

banderror commented on x-pack/plugins/security_solution/public/common/components/ml_popover/ml_modules.tsx:
```diff
@@ -11,8 +11,8 @@
 export const mlModules: string[] = [
-  'siem_cloudtrail',
-  'siem_packetbeat',
+  'ssecurity_cloudtrail',
```
Double `s` - probably a typo?

rylnd commented:
>  Could you double check if the renaming of the modules has any impact on your side?

@jpdjere thanks for the ping; I'm not certain on what the exact behavior is going to be here currently, but I can share some context/thoughts:

* Because ML Modules don't (currently) have a categorization field associated with them, we cannot programmatically filter for security modules in the same way that we can Jobs (which have a `group: security` field). That `mlModules` list is our allowlist for Security's jobs, so we'd need to add the "new" module names to that list in order for them to appear in our UI.
* My main concern is around how those old jobs are going to behave with this rename. I _believe_ we  mainly use the `job_id` for the purposes of identification, and so existing jobs will continue to show as installed despite the module rename, but we should verify that before this is released.

pheyos commented:
Hi @ajosh0504,

> I realized that there's some other siem references in the tests owing to the folders [here](https://github.com/elastic/kibana/tree/main/x-pack/test/functional/es_archives/ml) being named as siem_*. Is this sometime we want to modify in this PR as well?

The esArchives have been named close to the module ids to make it easier for developers to recognize / find them, so it would make sense to rename the archives as part of this PR. Note, tat this should also include updating the index name used by the data archive. Let me know if you need help with this or prefer to have this in a separate PR.

> Another thing I noted was references to moduleIDs containing versions (for example [here](https://github.com/elastic/kibana/blob/main/x-pack/test/api_integration/apis/ml/modules/get_module.ts#L30)), but those module IDs don't really exist [here](https://github.com/elastic/kibana/tree/main/x-pack/plugins/ml/server/models/data_recognizer/modules)?

Please note, that the directory names listed [here](https://github.com/elastic/kibana/tree/main/x-pack/plugins/ml/server/models/data_recognizer/modules) are not necessarily the module ids. The actual module ids are defined in the manifest files, se e.g. [here](https://github.com/elastic/kibana/blob/main/x-pack/plugins/ml/server/models/data_recognizer/modules/security_linux/manifest.json#L2).

ajosh0504 commented:
I've requested @pheyos to help with the esArchive naming update.

@rylnd I've added the new module names to the `mlModules` list as pointed out by @machadoum [earlier](https://github.com/elastic/kibana/pull/146760#issuecomment-1333728418). Also, since we aren't backporting these changes, I don't see why old jobs should be impacted. But yes, let me see if I can verify this.

pheyos commented:
@ajosh0504 the esArchive rename (including renaming of the contained indices) is done in 5e60cec.

ajosh0504 commented:
@pheyos Seeing some failing tests after your commit, although looking at the logs, I'm not sure if the failures are related to your changes. Could you take a look?

pheyos commented:
@ajosh0504 yes, the failures seem unrelated to the archive rename. Might be worth to comment `@elasticmachine merge upstream`.

ajosh0504 commented:
@elasticmachine merge upstream

peteharverson commented on x-pack/plugins/ml/server/models/data_recognizer/modules/security_cloudtrail/ml/rare_method_for_a_city.json:
```diff
@@ -29,6 +29,7 @@
   "custom_settings": {
-    "created_by": "ml-module-siem-cloudtrail"
+    "created_by": "ml-module-security-cloudtrail",
+    "security_app_display_name": "Unusual City For an AWS Command"
```
Nit - would use lowercase `for` here - which would be consistent with what you have for the username job.

peteharverson commented on x-pack/plugins/ml/server/models/data_recognizer/modules/security_cloudtrail/ml/rare_method_for_a_country.json:
```diff
@@ -29,6 +29,7 @@
   "custom_settings": {
-    "created_by": "ml-module-siem-cloudtrail"
+    "created_by": "ml-module-security-cloudtrail",
+    "security_app_display_name": "Unusual Country For an AWS Command"
```
Nit - would use lowercase `for` here - which would be consistent with what you have for the username job.

peteharverson commented on x-pack/plugins/ml/server/models/data_recognizer/modules/security_linux/ml/v3_linux_anomalous_process_all_hosts.json:
```diff
@@ -60,6 +60,7 @@
       }
-    ]
+    ],
+    "security_app_display_name": "Anomalous Process For a Linux Population"
```
Nit - would use lowercase `for` here - which would be consistent with what you have for some others job.

peteharverson commented on x-pack/plugins/ml/server/models/data_recognizer/modules/security_linux/ml/v3_rare_process_by_host_linux.json:
```diff
@@ -60,6 +60,7 @@
       }
-    ]
+    ],
+    "security_app_display_name": "Unusual Process For a Linux Host"
```
Nit - would use lowercase `for` here - which would be consistent with what you have for the username job.

peteharverson commented on x-pack/plugins/ml/server/models/data_recognizer/modules/security_network/ml/high_count_by_destination_country.json:
```diff
@@ -31,6 +31,7 @@
   "custom_settings": {
-    "created_by": "ml-module-security-network"
+    "created_by": "ml-module-security-network",
+    "security_app_display_name": "Spike in Network Traffic To a Country"
```
Nit - would use lowercase `to` here.

peteharverson commented on x-pack/plugins/ml/server/models/data_recognizer/modules/security_windows/ml/v3_windows_anomalous_process_all_hosts.json:
```diff
@@ -61,6 +61,7 @@
       }
-    ]
+    ],
+    "security_app_display_name": "Anomalous Process For a Windows Population"
```
Nit - would use lowercase `for` here - which would be consistent with what you have for other jobs.

peteharverson reviewed (approved):
ML changes LGTM. Just left some minor comments for the `security_app_display_name` values.

I was able to test creation of the jobs inside ML for:
- security_auth
- security_cloudtrail
- security_linux
- security_windows

Jobs were created successfully - example from the cloudtrail jobs:

<img width="1403" alt="image" src="https://user-images.githubusercontent.com/7405507/207638190-018ff5c3-034c-4ef0-b312-e2ce69c14903.png">

ajosh0504 commented:
@ryland I ran some tests locally to check the impact on existing ML jobs.

**Steps**
- Changed the `security_cloudtrail` module to how to was before i.e. `siem_cloudtrail` and containing `siem` references in the job configs, manifest etc.
- Enabled the Cloudtrail module in my local cluster. Ensured that I was seeing the old configurations- note that the `created_by` field in the screenshot says `ml-module-siem-cloudtrail`
<img width="1659" alt="Screen Shot 2022-12-14 at 2 49 17 PM" src="https://user-images.githubusercontent.com/30438249/207732948-37870f3e-8b3b-4c68-af69-a2ee04afa3f5.png">
- Changed the module id back to `security_cloudtrail` and remove other `siem` references.
- Noticed errors saying the job ids and datafeeds already exist.
<img width="1647" alt="Screen Shot 2022-12-14 at 2 51 50 PM" src="https://user-images.githubusercontent.com/30438249/207733513-be1a49a5-4674-4787-9ec1-aa15b0747143.png">

This means that updating the module id and removing other `siem` references from the job configurations will not impact old jobs.

spong commented:
@ajosh0504 @rylnd -- just now seeing this PR, and was also curious about the upgrade flow and UI functionality within the Security app with this module name change so I did a little testing as well.

Similar to @ajosh0504's test above, I checked out the commit before this landed 578d643 where the module was still named `siem_packetbeat`, setup packetbeat, then installed the jobs via the Security `ML Job Settings` UI, and then was able to successfully enable them as well. I then stopped kibana, checked out this branch, and verified the Security `ML Jobs Settings` UI functioned without issue. The previous running job was still enabled, and it was still possible to stop/start it or any of the other jobs within the same module. 

I was then curious if there would be any hiccups in the Security UI if needing to re-install any of the jobs, so I went to the ML UI, deleted one job (`packetbeat_rare_urls`), then went back to the Security UI and enabled that job and it successfully re-installed and enabled the job without issue/error 🎉. You can see in the image below the different `created_by` values as they differ between the module update:

<p align="center">
  <img width="500" src="https://user-images.githubusercontent.com/2946766/208206227-26e8c588-bd9b-4d6c-8991-1f6f998fe48b.png" />
</p>

So as you suspected @rylnd, since everything on the Security side is using `job_id` for identification (and those didn't change here), we're all good with regards to showing jobs and their relationship to any detection rules that may be active. Additionally (and where my concern was), the Security job management UI's still function for install/enable/disable when previous installs are present, and the Detection Rule Job selector has no issues displaying the jobs either.

Also, I went went ahead and created [this enhancement](https://github.com/elastic/kibana/issues/147722) for surfacing the new `security_app_display_name` field in our Security UI's. Thank you @ajosh0504 for adding this! 🙂

ajosh0504 commented:
@spong Thank you so much for the additional tests you ran. Glad that my changes don't break things at y'all's end. :)

Also, yay enhancements!
